contracts/ERC7432/ERC7432WrapperForERC4907.sol

```diff
@@ -0,0 +1,250 @@
+            _role.revocable,
+            _role.data
+        );
+    }
+
+    function revokeRole(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external override onlyUserRole(_roleId) {
+        address _wrappedTokenAddress = oriumWrapperManager.getWrappedTokenOf(_tokenAddress);
+        require(_wrappedTokenAddress != address(0), 'ERC7432WrapperForERC4907: token not supported');
+
+        address _recipient = IERC4907(_wrappedTokenAddress).userOf(_tokenId);
+        address _caller = _getApprovedCaller(_tokenAddress, _tokenId, _recipient);
+
+        // if caller is recipient, the role can be revoked regardless of its state
+        if (_caller != _recipient) {
+            // if caller is owner, the role can only be revoked if revocable or expired
+            require(
+                isRevocableRole[_tokenAddress][_tokenId] ||
+                    IERC4907(_wrappedTokenAddress).userExpires(_tokenId) < block.timestamp,
+                'ERC7432WrapperForERC4907: role is not revocable nor expired'
+            );
+        }
+
+        delete isRevocableRole[_tokenAddress][_tokenId];
+        IERC4907(_wrappedTokenAddress).setUser(_tokenId, address(0), uint64(0));
+        emit RoleRevoked(_tokenAddress, _tokenId, _roleId);
+    }
+
+    function setRoleApprovalForAll(address _tokenAddress, address _operator, bool _approved) external override {
+        tokenApprovals[msg.sender][_tokenAddress][_operator] = _approved;
+        emit RoleApprovalForAll(_tokenAddress, _operator, _approved);
+    }
+
+    /** ERC-7432 View Functions **/
+
+    function recipientOf(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (address recipient_) {
+        address _wrappedTokenAddress = oriumWrapperManager.getWrappedTokenOf(_tokenAddress);
+        if (_wrappedTokenAddress == address(0) || _roleId != USER_ROLE) {
+            return address(0);
+        }
+        return IERC4907(_wrappedTokenAddress).userOf(_tokenId);
+    }
+
+    function roleData(address, uint256, bytes32) external pure returns (bytes memory) {
+        return '';
+    }
+
+    function roleExpirationDate(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (uint64 expirationDate_) {
+        address _wrappedTokenAddress = oriumWrapperManager.getWrappedTokenOf(_tokenAddress);
+        if (_wrappedTokenAddress == address(0) || _roleId != USER_ROLE) {
+            return 0;
+        }
+        return uint64(IERC4907(_wrappedTokenAddress).userExpires(_tokenId));
+    }
+
+    function isRoleRevocable(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (bool revocable_) {
+        return
+            _roleId == USER_ROLE &&
+            isRevocableRole[_tokenAddress][_tokenId] &&
+            oriumWrapperManager.getWrappedTokenOf(_tokenAddress) != address(0);
```

Comment:
ERC-7432 users should always check the value of the expiration date, regardless of the value of revocable. I think it would be a waste of gas, imagine someone is building the following code:
```solidity
bool canBeRevoked = IERC7432(_tokenAddress).isRoleRevocable(_tokenAddress, _tokenId, _roleId) &&
    IERC7432(_tokenAddress).roleExpirationDate(_tokenAddress, _tokenId, _roleId) > block.timestamp;
```

In a nutshell, the `isRoleRevocable` would call `roleExpirationDate` under the hood, which sounds redundant.